### PR TITLE
Adsb support

### PR DESCRIPTION
**Pull request by ShikOfTheRa**
ADSB support for uAvionix pingRX tiny ADSB (and others).
OSD displays the nearest ADSB equipped aircraft.
OSD adsb data flashes as warning if within set distance for proximity warning of other aircraft

Wiki page prepared with full details:
[](https://github.com/ShikOfTheRa/inav/wiki/ADSB)

(updated and with missing files added from original PR:  #6192)

GUI PR #1294

### CODE REVIEW

File: src/main/io/adsb.c [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+void GPS_distance_cm_bearing(int32_t currentLat1, int32_t currentLon1, int32_t destinationLat2, int32_t destinationLon2, uint32_t *dist, int32_t *bearing)
+{
+  #define DISTANCE_BETWEEN_TWO_LONGITUDE_POINTS_AT_EQUATOR    1.113195f  // MagicEarthNumber from APM
+  float GPS_scaleLonDown = cos_approx((fabsf((float)gpsSol.llh.lat) / 10000000.0f) * 0.0174532925f);
+  const float dLat = destinationLat2 - currentLat1; // difference of latitude in 1/10 000 000 degrees
+  const float dLon = (float)(destinationLon2 - currentLon1) * GPS_scaleLonDown;
+
+  *dist = sqrtf(sq(dLat) + sq(dLon)) * DISTANCE_BETWEEN_TWO_LONGITUDE_POINTS_AT_EQUATOR;
+  *bearing = 9000.0f + RADIANS_TO_CENTIDEGREES(atan2_approx(-dLat, dLon));      // Convert the output radians to 100xdeg
+  if (*bearing < 0){
```
Comment (shellixyz) [inline src/main/io/adsb.c]: You could probably use wrap_36000 there

File: src/main/io/adsb.c [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    *bearing += 36000;
+  }
+};
+
+void adsbNewVehicle(uint32_t avicao, int32_t avlat, int32_t avlon, int32_t avalt)
+{
+    uint32_t avdist; int32_t avdir; uint8_t avupdate = 1;
```
Comment (shellixyz) [inline src/main/io/adsb.c]: One line for each variable declaration please

File: src/main/io/adsb.c [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+#include "math.h"
+#include "common/time.h"
+#include "drivers/time.h"
+
+adsbVehicle_t adsb;
+static timeUs_t lastADSB = 0;
+
+void GPS_distance_cm_bearing(int32_t currentLat1, int32_t currentLon1, int32_t destinationLat2, int32_t destinationLon2, uint32_t *dist, int32_t *bearing)
+{
+  #define DISTANCE_BETWEEN_TWO_LONGITUDE_POINTS_AT_EQUATOR    1.113195f  // MagicEarthNumber from APM
```
Comment (shellixyz) [inline src/main/io/adsb.c]: This should be defined elsewhere. It is already defined in `navigation_private.h`. Perhaps it should be moved in `navigation.h`

File: src/main/io/adsb.c [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+void adsbNewVehicle(uint32_t avicao, int32_t avlat, int32_t avlon, int32_t avalt)
+{
+    uint32_t avdist; int32_t avdir; uint8_t avupdate = 1; 
+    if (STATE(GPS_FIX) && gpsSol.numSat >= 5){
+      GPS_distance_cm_bearing(gpsSol.llh.lat, gpsSol.llh.lon, avlat, avlon, &avdist, &avdir); 
+    }
+    avdist /= 100; avdir /= 100;
+#if defined(USE_NAV)
+    avalt -= getEstimatedActualPosition(Z)+GPS_home.alt;
```
Comment (shellixyz) [inline src/main/io/adsb.c]: Space around `+` for readability

File: src/main/io/adsb.c [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  *bearing = 9000.0f + RADIANS_TO_CENTIDEGREES(atan2_approx(-dLat, dLon));      // Convert the output radians to 100xdeg
+  if (*bearing < 0){
+    *bearing += 36000;
+  }
+};
+
+void adsbNewVehicle(uint32_t avicao, int32_t avlat, int32_t avlon, int32_t avalt)
+{
+    uint32_t avdist; int32_t avdir; uint8_t avupdate = 1; 
+    if (STATE(GPS_FIX) && gpsSol.numSat >= 5){
```
Comment (shellixyz) [inline src/main/io/adsb.c]: Space between `)` and `{`

File: src/main/io/adsb.c [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+      adsb.icao = avicao;
+      adsb.dist = avdist;       
+      adsb.alt = avalt; 
+      adsb.dir = avdir; 
+      adsb.ttl = 10;    // 10 secs default timeout    
+    }   
+};
+
+void adsbexpiry(){
+  uint64_t currentTimeUs = micros();
```
Comment (shellixyz) [inline src/main/io/adsb.c]: Please use timeUs_t instead of uint64_t

File: src/main/io/adsb.c [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+      adsb.ttl--;
+    }
+    else{
+      adsb.icao = 0;
+      adsb.dist = 0;       
+      adsb.alt = 0; 
+      adsb.dir = 0;       
+    }
+  }
+};
```
Comment (shellixyz) [inline src/main/io/adsb.c]: No need for `;` there

File: src/main/io/osd.c [hunk 8]
```diff
@@ -1619,7 +1620,35 @@ static bool osdDrawSingleElement(uint8_t item)
             osdFormatCentiNumber(&buff[2], centiHDOP, 0, 1, 0, 2);
             break;
         }
-
+#if defined(USE_TELEMETRY) && defined(USE_TELEMETRY_MAVLINK)
+    case OSD_ADSB:
+        {         
+            static uint8_t adsblen=1;
+            for (int i = 0; i <= adsblen; i++) {
```
Comment (shellixyz) [inline src/main/io/osd.c]: Please use memset instead of a for

File: src/main/io/osd.c [hunk 9]
```diff
@@ -1619,7 +1620,35 @@ static bool osdDrawSingleElement(uint8_t item)
...
+                buff[i] = SYM_BLANK;
+            }
+            buff[adsblen]='\0';
```
Comment (shellixyz) [inline src/main/io/osd.c]: Spaces around `=`

File: src/main/io/osd.c [hunk 10]
```diff
@@ -1619,7 +1620,35 @@ static bool osdDrawSingleElement(uint8_t item)
...
+            displayWrite(osdDisplayPort, elemPosX, elemPosY, buff); // clear any previous chars because variable element size
+            adsbexpiry();
+            adsblen=1;
+            buff[0] = SYM_ADSB;   
+            if ((adsb.dist > 0) && (adsb.dist < osdConfig()->adsb_range)){
+                osdFormatDistanceStr(&buff[1], adsb.dist*100);
```
Comment (shellixyz) [inline src/main/io/osd.c]: Please replace `&buff[1]` with `buff + 1`

File: src/main/io/osd.c [hunk 11]
```diff
@@ -1619,7 +1620,35 @@ static bool osdDrawSingleElement(uint8_t item)
...
+                adsblen = strlen(buff);
+                int dir = osdGetHeadingAngle(adsb.dir + 11);
+                unsigned arrowOffset = dir * 2 / 45;
+                buff[adsblen-1]=SYM_ARROW_UP + arrowOffset;
+                osdFormatDistanceStr(&buff[adsblen], adsb.alt*100);
```
Comment (shellixyz) [inline src/main/io/osd.c]: `&buff[adsblen]` -> `buff + adsblen`

File: src/main/fc/settings.yaml [hunk 12]
```diff
@@ -3211,6 +3210,18 @@ groups:
         field: sidebar_scroll_arrows
         type: bool
         default_value: OFF
+      - name: osd_adsb_range
+        description: "Distance outside which adsb aircraft are NOT displayed. In meters"
+        default_value: 20000
+        field: adsb_range
+        min: 0
```
Comment (shellixyz) [inline src/main/fc/settings.yaml]: Default min is 0 so this is not needed but it won't hurt

File: src/main/fc/settings.yaml [hunk 13]
```diff
@@ -3211,6 +3210,18 @@ groups:
         field: sidebar_scroll_arrows
         type: bool
         default_value: OFF
+      - name: osd_adsb_range
+        description: "Distance outside which adsb aircraft are NOT displayed. In meters"
```
Comment (shellixyz) [inline src/main/fc/settings.yaml]: ADSB should be uppercased

**Review by shellixyz — CHANGES_REQUESTED** [review]
Please fix the formatting:
- 4 spaces indentation
- space between `)` and `{`, around math operators and `=`
- choose between snake case and camel case for function and argument names, please don't use both or full lowercase

File: src/main/io/adsb.c (hunk 1 above)
Comment (ShikOfTheRa) [reply src/main/io/adsb.c]: cool - will get all these done first thing tomorrow

Comment (ShikOfTheRa) [reply src/main/io/adsb.c]: Updated

File: src/main/io/adsb.c (hunk 2 above)
Comment (ShikOfTheRa) [reply src/main/io/adsb.c]: Updated

File: src/main/io/adsb.c (hunk 3 above)
Comment (ShikOfTheRa) [reply src/main/io/adsb.c]: If OK I would like to leave here - Once I have fully understood the navigation code I  think we can remove and migrate to use existing - but it is quite complex and will take some time..

File: src/main/io/adsb.c (hunk 4 above)
Comment (ShikOfTheRa) [reply src/main/io/adsb.c]: done

File: src/main/io/adsb.c (hunk 5 above)
Comment (ShikOfTheRa) [reply src/main/io/adsb.c]: done

File: src/main/io/adsb.c (hunk 6 above)
Comment (ShikOfTheRa) [reply src/main/io/adsb.c]: done

File: src/main/io/adsb.c (hunk 7 above)
Comment (ShikOfTheRa) [reply src/main/io/adsb.c]: done

File: src/main/io/osd.c (hunk 8 above)
Comment (ShikOfTheRa) [reply src/main/io/osd.c]: done

File: src/main/io/osd.c (hunk 9 above)
Comment (ShikOfTheRa) [reply src/main/io/osd.c]: done

File: src/main/io/osd.c (hunk 11 above)
Comment (ShikOfTheRa) [reply src/main/io/osd.c]: done

File: src/main/fc/settings.yaml (hunk 12 above)
Comment (ShikOfTheRa) [reply src/main/fc/settings.yaml]: noted. done

File: src/main/fc/settings.yaml (hunk 13 above)
Comment (ShikOfTheRa) [reply src/main/fc/settings.yaml]: done

**Pull request closed without merging**